e2e/playground.py

```diff
@@ -0,0 +1,2052 @@
+
+
+class _InputWithContainer(_InputBase):
+    loc_container: Locator
+
+    def __init__(
+        self,
+        page: Page,
+        *,
+        id: str,
+        loc: InitLocator,
+        loc_container: InitLocator = "div.shiny-input-container",
+    ):
+
+        loc_is_str = isinstance(loc, str)
+        loc_container_is_str = isinstance(loc_container, str)
+
+        if loc_is_str and loc_container_is_str:
+            loc_container = page.locator(loc_container).filter(
+                # `page.locator(loc)` is executed from within `loc_container`
+                has=page.locator(loc)
+            )
+
+            loc = loc_container.locator(loc)
+        elif not loc_is_str and not loc_container_is_str:
+            ...  # Do nothing; Use values as is
+        else:
+            raise ValueError(
+                "`loc` and `loc_container` must both be strings or both be Locators"
+            )
+
+        super().__init__(
+            page,
+            id=id,
+            loc=loc,
+        )
+        self.loc_container = loc_container
+
+
+class _InputWithLabel(_InputWithContainer):
+    def __init__(
+        self,
+        page: Page,
+        *,
+        id: str,
+        loc: InitLocator,
+        loc_container: InitLocator = "div.shiny-input-container",
+        loc_label: InitLocator = "label",
+    ):
+        super().__init__(
+            page,
+            id=id,
+            loc_container=loc_container,
+            loc=loc,
+        )
+
+        if isinstance(loc_label, str):
+            loc_label = self.loc_container.locator(loc_label)
+        self.loc_label = loc_label
+
+    def value_label(self, *, timeout: Timeout = None) -> typing.Union[str, None]:
+        return self.loc_label.text_content(timeout=timeout)
+
+    def expect_label_to_have_text(self, value: str, *, timeout: Timeout = None):
+        playwright_expect(self.loc_label).to_have_text(value, timeout=timeout)
+
+
+# Class definitions
+# * Fields
+#   * Try to mirror playwright as much as possible.
+#   * There are no properties, only methods; This allows for timeout values to be passed through and for complex methods.
+#     * Locators will stay as properties
+#   * Don't sub-class. For now, use `_` separatation and use `loc` or `value` as a prefix
+# * Approach
+#   * Use locators / playwright_expect as much as possible
+#     * It should not be necessary to use `assert` directly.
+#     * MUST wait for `Locator`s to do their job
+#   * Provide `value` methods as a convenience
+
+
+class _WidthLoc:
+    def value_width(
+        self: _InputBaseP,
+        *,
+        timeout: Timeout = None,
+    ) -> typing.Union[str, None]:
+        return str_attr(self.loc, "width", timeout=timeout)
+
+    def expect_width_to_have_value(
+        self: _InputBaseP,
+        value: typing.Union[AttrValue, None],
+        *,
+        timeout: Timeout = None,
+    ):
+        expect_attr(self.loc, "width", value=value, timeout=timeout)
+
+
+class _WidthContainer:
```

Comment:
Can the mixin classes inherit from abstract meta classes to prevent init and property values?